### PR TITLE
JAMES-3434 Fix EmailSubmission/set onSuccessUpdateEmail property

### PR DESCRIPTION
**Pull request by chibenwa**
Re-reading https://jmap.io/spec-mail.html#emailsubmissionset

```
onSuccessUpdateEmail: Id[PatchObject]|null A map of EmailSubmission id
to an object containing properties to update on the Email object
referenced by the EmailSubmission if the create/update/destroy
succeeds. (For references to EmailSubmissions created in the same
“/set” invocation, this is equivalent to a creation-reference, so the
id will be the creation id prefixed with a #.)
```

We are currently using a message id, not an EmailSubmission id...

CF a bug report opened on the Linagora fork: https://github.com/linagora/james-project/issues/4310

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 1]
```diff
@@ -126,6 +129,29 @@ class EmailSubmissionSetMethod @Inject()(serializer: EmailSubmissionSetSerialize
         case _ => None
       }
       .toMap
+
+    def resolveMessageId(creationId: EmailSubmissionCreationId): Either[IllegalArgumentException, MessageId] = {
+      if (creationId.startsWith("#")) {
+        val realId = creationId.substring(1)
+        val validatedId: Either[String, EmailSubmissionCreationId] = refineV[IdConstraint](realId)
+        validatedId
+          .left.map(s => new IllegalArgumentException(s))
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:138]: I don't understand why you need that line?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala [hunk 2]
```diff
@@ -185,12 +215,14 @@ class EmailSubmissionSetMethod @Inject()(serializer: EmailSubmissionSetSerialize
...
+        case (creationResponse, messageId) =>
+          recordCreationIdInProcessingContext(emailSubmissionCreationId, processingContext, creationResponse.id)
+            .map(context => (creationResponse, messageId, context))
+      }
       .fold(e => (CreationFailure(emailSubmissionCreationId, e), processingContext),
-        creationResponseWithUpdatedContext => {
-          (CreationSuccess(emailSubmissionCreationId, creationResponseWithUpdatedContext._1), creationResponseWithUpdatedContext._2)
-        })
+        creation =>
+          (CreationSuccess(emailSubmissionCreationId, creation._1, creation._2), creation._3))
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:227]: extra useless parenthesis?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala [hunk 3]
```diff
@@ -63,6 +64,33 @@ case class EmailSubmissionSetRequest(accountId: AccountId,
...
+      .sequence
+      .map(_ => this)
+  }
+
+  private def validate(creationId: EmailSubmissionCreationId, supportedCreationIds: List[EmailSubmissionCreationId]): Either[IllegalArgumentException, EmailSubmissionCreationId] = {
+    if (creationId.startsWith("#")) {
+      val realId = creationId.substring(1)
+      val validatedId: Either[String, EmailSubmissionCreationId] = refineV[IdConstraint](realId)
+      validatedId
+        .left.map(s => new IllegalArgumentException(s))
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala:105]: is this line necessary?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 4]
```diff
@@ -1165,6 +1165,135 @@ trait EmailSubmissionSetMethodContract {
                    |}""".stripMargin)
   }
 
+
+  @Test
+  def test(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:1170]: test?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala [hunk 5]
```diff
@@ -1165,6 +1165,135 @@ trait EmailSubmissionSetMethodContract {
...
+                    |                "description": "notFound cannot be retrieved as storage for EmailSubmission is not yet implemented"
+                    |            },
+                    |            "c1"
+                    |        ]
+                    |    ]
+                    |}""".stripMargin)
+  }
+
+  @Test
+  def test2(server: GuiceJamesServer): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailSubmissionSetMethodContract.scala:1234]: test2?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala (hunk 3 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailSubmissionSet.scala:105]: Yes, as we need to transform `Either[String, X]` into `Either[IllegalArgumentException, X]`

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:138]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:227]: No, it is required to have a Tuple2. The syntax is `(a, b)`

Comment (jeantil) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:227]: alternative syntax for tuple2 `a->b`

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala:227]: Thanks for the tip!

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailSubmissionSetMethod.scala]: Only Tuple2 right?

**Review by Arsnael — APPROVED** [review]

**Pull request merged**